### PR TITLE
fix bugs and improvement for mo_ctl supported traces

### PR DESCRIPTION
**Pull request by gouhongshen**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:
In the previous PR #11607, we supported mo_ctl controlled trace. In that implementation, span kind is used to help control these special traces enable or disable, however, these special kinds could be passed to its children trace or overwritten by its parent trace. 

these problems have a negative impact on the accuracy of collected statistics and trace performance.

this PR has fixed them.

and:

a new span to observe s3GetObject has been added and supports `all` to simplify the list of cn uuids.

### CODE REVIEW

**Comment by matrix-meow** [comment]
@gouhongshen Thanks for your contributions!

## Review:

### Title
The title of the pull request is "fix bugs". While it is concise, it does not provide specific information about the bugs being fixed. It would be helpful to have a more descriptive title that summarizes the changes made in the pull request.

### Body
The body of the pull request provides some context about the previous implementation and the problems that were identified. However, it lacks specific details about the bugs and the impact they had on the codebase. It would be beneficial to include more information about the specific issues that were fixed and how they were affecting the accuracy of collected statistics and trace performance.

### Changes
1. In `pkg/util/trace/config.go`, a new method `HasMOCtledKind()` is added to the `SpanContext` struct. This method checks if the `SpanContext` has a special kind that is controlled by `mo_ctl`.
   - The method implementation looks fine and is straightforward.

2. In `pkg/util/trace/config.go`, the `SpanKind` constants `SpanKindS3FSVis` and `SpanKindLocalFSVis` are changed from untyped constants to typed constants.
   - This change improves type safety and clarity in the code.

3. In `pkg/util/trace/impl/motrace/mo_trace.go`, the `Start` method is modified to handle the special kinds of spans controlled by `mo_ctl`.
   - The code logic for handling the special kinds of spans looks correct and addresses the problem mentioned in the pull request body.

4. In `pkg/util/trace/impl/motrace/mo_trace_bench_test.go`, a new benchmark test is added to compare the performance of the `IsEnable` and `Start` methods with and without options.
   - The benchmark test looks fine and provides useful performance comparison.

5. In `pkg/util/trace/impl/motrace/mo_trace_test.go`, two new test cases are added to test the behavior of the special kinds of spans controlled by `mo_ctl`.
   - The test cases look fine and cover the expected behavior.

### Suggestions
1. Provide a more descriptive title that summarizes the changes made in the pull request.

2. Add more specific details in the pull request body about the bugs that were fixed and the impact they had on the codebase.

3. Consider adding comments to the code changes to provide more clarity and explanation for future developers.

4. Consider adding unit tests to cover the new code changes and ensure their correctness.

5. Consider optimizing the code changes further if possible, based on performance analysis and profiling.

Overall, the changes made in the pull request address the identified problems and improve the codebase. However, providing more specific details and adding comments and tests would enhance the quality and maintainability of the code.

File: pkg/util/trace/impl/motrace/mo_trace.go [hunk 1]
```diff
@@ -91,9 +91,19 @@ func (t *MOTracer) Start(ctx context.Context, name string, opts ...trace.SpanSta
...
 	}
+	// in previous implementation, if parent span exists, it's span kind will
+	// be overwritten by its parent whatever its kind has been set or not by `opts`
+	//
+	// however, if there exists a special kind (controlled by mo_ctl) has been set in `span.init`, or
+	// parent span has a special kind, we hope:
+	// 	  1. this kind won't be overwritten by it parent span kind if it exists
+	//	  2. this special kind won't continue to pass down to its children span
+	if !span.HasMOCtledKind() && !psc.HasMOCtledKind() {
+		span.Kind = psc.Kind
```
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/mo_trace.go:105]: 这里没有一定要继承 parent.kind，如果你需要不继承，这里可以不用保留原有逻辑。

Comment (gouhongshen) [reply pkg/util/trace/impl/motrace/mo_trace.go]: 好的，那我就去掉这部分代码

File: pkg/util/trace/impl/motrace/mo_trace.go [hunk 2]
```diff
@@ -91,7 +91,7 @@ func (t *MOTracer) Start(ctx context.Context, name string, opts ...trace.SpanSta
 		span.TraceID, span.SpanID = t.provider.idGenerator.NewIDs()
 		span.Parent = trace.NoopSpan{}
 	} else {
-		span.TraceID, span.SpanID, span.Kind = psc.TraceID, t.provider.idGenerator.NewSpanID(), psc.Kind
+		span.TraceID, span.SpanID = psc.TraceID, t.provider.idGenerator.NewSpanID()
```
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/mo_trace.go:94]: LGTM

**Review by xzxiong — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

**Pull request merged**